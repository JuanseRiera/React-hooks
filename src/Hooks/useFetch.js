import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
  const isMounted = useRef(true);
  const [state, setstate] = useState({
    loading: true,
    data: null,
    error: null,
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    if (url) {
      setstate({
        loading: true,
        data: null,
        error: null,
      });
      fetch(url)
        .then((resp) => resp.json())
        .then((respuesta) => {
          if (isMounted.current) {
            setstate({
              loading: false,
              data: respuesta,
              error: null,
            });
          } else {
            console.log("No se llamo el setstate");
          }
        });
    } else {
      setstate({
        loading: false,
        data: null,
        error: null,
      });
    }
  }, [url]);

  return state;
};
