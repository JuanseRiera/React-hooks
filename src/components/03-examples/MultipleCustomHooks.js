import React from "react";
import { useCounter } from "../../Hooks/useCounter";
import { useFetch } from "../../Hooks/useFetch";
import "./MultipleCustomHooks.css";

export const MultipleCustomHooks = () => {
  const { counter, incrementar } = useCounter(1);
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { quote, author } = !!data && data[0];

  return (
    <>
      <h1>Breaking Bad quotes</h1>
      <hr />

      {loading ? (
        <div className="mt-5 alert alert-info">Cargando informaci√≥n...</div>
      ) : (
        <>
          <div className="mt-5">
            <figure className="text-center">
              <blockquote className="blockquote">
                <p id="quote">{quote}</p>
              </blockquote>
              <figcaption className="blockquote-footer">{author}</figcaption>
            </figure>
          </div>
          <div className="mt-3 text-end">
            <button
              className="btn btn-info"
              onClick={() => {
                incrementar(1);
              }}
            >
              Siguiente quote
            </button>
          </div>
        </>
      )}
    </>
  );
};
