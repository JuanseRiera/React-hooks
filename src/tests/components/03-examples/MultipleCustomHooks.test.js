import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../Hooks/useFetch";

jest.mock("../../../Hooks/useFetch");

describe("Pruebas del componente multiple custom hooks", () => {
  test("Deberia cargar correctamente", () => {
    useFetch.mockReturnValue({
      loading: true,
      data: null,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Deberia mostrar como autor juanse y como frase hola mundo", () => {
    useFetch.mockReturnValue({
      loading: false,
      data: [
        {
          author: "juanse",
          quote: "Hola mundo",
        },
      ],
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find("#quote").text().trim()).toBe("Hola mundo");
    expect(wrapper.find("figcaption").text().trim()).toBe("juanse");
  });
});
