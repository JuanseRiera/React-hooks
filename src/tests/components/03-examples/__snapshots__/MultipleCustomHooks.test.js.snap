// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente multiple custom hooks Deberia cargar correctamente 1`] = `
<Fragment>
  <h1>
    Breaking Bad quotes
  </h1>
  <hr />
  <div
    className="mt-5 alert alert-info"
  >
    Cargando informaci√≥n...
  </div>
</Fragment>
`;
